File: core/src/com/oni/donjon/map/TileType.java
package com.oni.donjon.map;

import com.badlogic.gdx.graphics.Texture;
import com.oni.donjon.screen.GameScreen;

import java.util.HashMap;

/**
 * @author Daniel Chesters (on 22/05/14).
 */
public enum TileType {
    WALL("textures/wall.png", GameScreen.WALL_BIT),
    DOOR_CLOSE("textures/door_close.png", GameScreen.WALL_BIT),
    DOOR_OPEN("textures/door_open.png", GameScreen.NOTHING_BIT),
    GROUND("textures/ground.png", GameScreen.NOTHING_BIT),
    STAIR_UP("textures/stair_up.png", GameScreen.NOTHING_BIT),
    STAIR_DOWN("textures/stair_down.png", GameScreen.NOTHING_BIT);

    private static final java.util.Map<String, Texture> cacheTexture = new HashMap<>();

    private String texturePath;
    private short categoryBits;

    TileType(String texturePath, short categoryBits) {
        this.texturePath = texturePath;
        this.categoryBits = categoryBits;
    }

    public Texture getTexture() {
        if (!cacheTexture.containsKey(texturePath)) {
            cacheTexture.put(texturePath, new Texture(texturePath));
        }
        return cacheTexture.get(texturePath);
    }

    public short getCategoryBits() {
        return categoryBits;
    }
}

File: core/src/main/kotlin/com/oni/donjon/map/TileType.kt
package com.oni.donjon.map

import com.badlogic.gdx.graphics.Texture
import com.oni.donjon.screen.GameScreen
import ktx.collections.gdxMapOf
import ktx.collections.set

/**
 * @author Daniel Chesters (on 22/05/14).
 */
enum class TileType constructor(private val texturePath: String, val categoryBits: Short) {
    WALL("textures/wall.png", GameScreen.WALL_BIT),
    DOOR_CLOSE("textures/door_close.png", GameScreen.WALL_BIT),
    DOOR_OPEN("textures/door_open.png", GameScreen.NOTHING_BIT),
    GROUND("textures/ground.png", GameScreen.NOTHING_BIT),
    STAIR_UP("textures/stair_up.png", GameScreen.NOTHING_BIT),
    STAIR_DOWN("textures/stair_down.png", GameScreen.NOTHING_BIT);

    val texture: Texture?
        get() {
            if (!cacheTexture.containsKey(texturePath)) {
                cacheTexture[texturePath] = Texture(texturePath)
            }
            return cacheTexture[texturePath]
        }

    companion object {
        private val cacheTexture = gdxMapOf<String, Texture>()
    }
}
